sstables/storage.cc

```diff
@@ -319,26 +355,28 @@ future<> filesystem_storage::create_links_common(const sstable& sst, sstring dst
     // TemporaryTOC is always first, TOC is always last
     auto dst = sstable::filename(dst_dir, sst._schema->ks_name(), sst._schema->cf_name(), sst._version, generation, sst._format, component_type::TemporaryTOC);
     co_await sst.sstable_write_io_check(idempotent_link_file, sst.filename(component_type::TOC), std::move(dst));
-    co_await sst.sstable_write_io_check(sync_directory, dst_dir);
+    auto dir_f = co_await open_checked_directory(sst._write_error_handler, dst_dir);
```

Comment:
I think it would be more elegant to reuse open_directory for dst_dir as well, even though it's for a single use.
(and `co_await open(sst._write_error_handler)` it here right after construction)
This way we would use one proven way of doing things.  It shouldn't have much overhead over the open coded implementation.

Reply:
@bhalevy 
> I think it would be more elegant to reuse open_directory for dst_dir as well, even though it's for a single use. [...] This way we would use one proven way of doing things. It shouldn't have much overhead over the open coded implementation.

OK, I'll investigate that, for a potential v7.

> (and `co_await open(sst._write_error_handler)` it here right after construction)

Again, IMO that should be unnecessary; the first sync will perform the (error-checked) open. There seems to be no other use for the open than enabling sync'ing, so making it explicit looks redundant to me.

Reply:
> @bhalevy
> 
> > I think it would be more elegant to reuse open_directory for dst_dir as well, even though it's for a single use. [...] This way we would use one proven way of doing things. It shouldn't have much overhead over the open coded implementation.
> 
> OK, I'll investigate that, for a potential v7.

It turns out that there's a small dependency glitch / challenge for this. Namely, in order to use `opened_directory` in patch#4 (_sstables/storage: synch "dst_dir" more leanly in create_links_common()_), and to do so _temporarily_ -- for replacing the current, also tempoary, _dir_f_ object --, the `opened_directory::close()` method must be available. However, in version 6 of the series, `opened_directory::close()` is only introduced in patch#6 (_sstables/storage: close previous directory asynchronously upon dir change_).

I could of course move that four-liner hunk, introducing `opened_directory::close()`, from patch#6 to patch#4, technically speaking. But I don't like the idea. That code movement feels very much "bolted on". `create_links_common()` is not a first-class user of `opened_directory` (because it does not work with the sstable base directory). `create_links_common()` is more like an "opportunistic" user. Thus introducing `opened_directory::close()` just for its sake feels unjustified.

Instead, I'm going to move patch#4 to the end of the series (this will also match its "opportunistic user" nature). At that point, `opened_directory::close()` will be available anyway (for stronger reasons!), and then `create_links_common()` can opportunistically capitalize on the `opened_directory` class, including the `close()` method thereof.

Reply:
Sounds good (I think :))